keps/sig-api-machinery/20190923-extract-code-generator.md

```diff
@@ -0,0 +1,151 @@
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+k8s.io/code-generator is logically a separate component and one that we have structured as a staging repo before extraction.
+After we make the k8s.io/code-generator authoritative, we will vendor the repo into k8s.io/kubernetes to perform our generation.
+
+## Motivation
+
+The k8s.io/code-generator is externally valuable and there is no compelling reason to keep it in tree.
+We want to become a standard consumer of generator like all the others. 
```

Comment:
There is two compelling reasons for keeping it in tree, both which we are suffering from with kube-openapi:

1. vendoring in a new code-generator version will include multiple upstream changes. The person doing the revendoring will usually not be competent enough to judge whether the changes of the regenerated code are good or not.
2. test coverage of code-generator changes will dramatically be reduced. Before kube served as the test fixture, after this KEP it cannot anymore, or only with the additional work of making a "proof PR against kube".

What is the value of splitting it? "no compelling reason to keep it in tree" is not value by itself. "Reduced k/k size" might be.

Reply:
Both of those claims come down to a claim of a lack of discipline and not treating a repo like a project in its own right.  People use these generators.  You would not say that kube-builder suffered because it was developed out of tree.  In fact, I'd claim that it succeeded in large part because it was separated.

I'd like to see our generators given the same chance. And as I noted, if there is a problem or doubts with generation, the k/k repo will not be injured.  I would not let fear of change stagnate generation.

Reply:
Mitigation below looks good. tl/dr: fake-proof-PRs in k/k for non-trivial changes (we don't have too many nowadays) will do.

Reply:
resolved.